### ISSUE TITLE
Add support for `ghci-ng` in the Haskell layer

### ISSUE THREAD

**Issue by bjarkevad** [syl20bnr/spacemacs#346]
There's a fork of `ghci`, [ghci-ng](https://github.com/chrisdone/ghci-ng) which seems to add some interesting features. It's already supported in `haskell-mode`, so it would be easy to add support for it in Spacemacs' Haskell layer.
I suggest doing so by adding a flag, e.g. `haskell/use-ghci-ng` which swaps out the old `ghci` with `ghci-ng` as well as add any relevant keybindings (if any), if the user wishes to do so. That way no existing configs will be broken.

Any objections to this proposal? :)

**Comment by bjarkevad** [comment]
`ghci-ng` is currently not up-to-date when installed from hackage, which means that users would have to compile the package from source. This is very easy, but still a bit of an inconvinience, so I don't think we should use `ghci-ng` just yet. @syl20bnr, should I close the issue or keep it open?

**Comment by syl20bnr** [comment]
@bjarkevad you can close it and open a new one when `ghci-ng

**Comment by angerman** [comment]
Couldn't we have a `haskell-ng` layer for those who do have `ghci-ng`? As I understand, `ghci-ng` needs to be installed from source, due to hackage not supporting `-dynamic`. (See https://github.com/chrisdone/ghci-ng/issues/11)

**Comment by bjarkevad** [comment]
@angerman :+1: that's not a bad idea :)

**Comment by meditans** [comment]
@angerman supporting that!

**Comment by angerman** [comment]
@bjarkevad, @syl20bnr, @meditans: I spoke to @dcoutts the other day on irc, and was told that a hackage update might likely happen over the weekend (see haskell/hackage-server#308). Which would make `ghci-ng` available on hackage. I'm still on the fence, if having a `haskell` layer and a `haskell-ng` layer would be the better solution altogether, such that people can choose which setup they prefer?

**Comment by meditans** [comment]
@angerman, imho having two layers (one in the spirit of the current `haskell` layer, with ghc-mod and company-ghc, and the other, `haskell-ng`, more in the spirit of chrisdone's tooling, with ghci-ng, hindent and the other stuff) is the better way to go: in this way we are not forcing users to change their haskell tooling when switching to spacemacs, we offer them a choice to explore different tooling setups, and we are more resilient to the absence of one specific tool (if a user can't install ghc-mod, for example, [see https://github.com/kazu-yamamoto/ghc-mod/issues/417 which prevented me from using ghc-mod for a couple weeks], we can always direct him towards the ghci-ng solution.)

**tl;dr** I prefer the solution with two layers

**Comment by angerman** [comment]
Sounds reasonable. @bjarkevad is your layer still working? Could you adjust it to be `haskell-ng`? @syl20bnr would you be ok with having two haskell layers? `haskell` and `haskell-ng`?

**Comment by syl20bnr** [comment]
I prefer one layer with a layer variable to choose between the two setup.
Two layers for the same language is very confusing for the user and they could even activate both at the same time which would lead to undefined behavior or to a subjective choice (which layer has precedence on the other ? Etc...).

**Comment by bjarkevad** [comment]
@angerman I'll try to get it running this weekend. It's a rather simple fix iirc, so feel free to give it a shot yourself as well :-)

**Comment by bjarkevad** [comment]
I'm going to reopen the issue now, since it seems to be relevant again.

**Comment by angerman** [comment]
Can you roll that into the `haskell` layer with a variable, like suggested?

**Comment by bjarkevad** [comment]
@angerman I'll give it a shot

**Comment by angerman** [comment]
awesome!

**Comment by bjarkevad** [comment]
@anger see #655, seems to work really well! I'd be very grateful if you'd test it out and give me any feedback you might have.

**Comment by meditans** [comment]
@bjarkevad works for me, indeed. A minor preliminary feedback could be that in the documentation you could add a header for ghci-ng usage (currently the instructions are under OSX), so that it ends up in the toc.

**Comment by bjarkevad** [comment]
@meditans Thanks, I might not have checked the README thoroughly..

**Comment by syl20bnr** [comment]
Actually @bjarkevad did put a header on it but it was formatted with `***` instead of `###` so it is my fault (bad review :-)).

I just fixed it.

**Comment by angerman** [comment]
@bjarkevad brilliant!

Are we going to add more of the @chrisdone tooling (as @meditans called it) into the haskell layer in this fashion as well?

**Comment by bjarkevad** [comment]
I've been using org for too long it seems ;-)

**Comment by meditans** [comment]
@angerman @bjarkevad I'd be curious to see how well structured-haskell-mode could be adapted to vim-like keybindings (maybe aiming to an experience similar to the lisp-mode in spacemacs).

**Comment by bjarkevad** [comment]
@meditans I'll have a look

**Comment by bjarkevad** [comment]
So this is implemented now, closing. @meditans see #676 for SHM progress
